### PR TITLE
JS: Recognize object transformations to exported values when looking for library inputs

### PR DESCRIPTION
**Pull request by erik-krogh**
And a drive-by model of `async-execute`.  

Gets a TP/TN for CVE-2021-3190 

I'll do a run-on-all together with https://github.com/github/codeql/pull/5439 before merging.

### CODE REVIEW

**Comment by erik-krogh** [comment]
[Evaluation looks OK](https://github.com/dsp-testing/erik-krogh-dca/tree/run/asExec-nightly-LibInputs/reports).  

I couldn't replicate the performance regression locally. So I think it might be spurious. 
(The evaluation is only on 2 queries). 

(Regarging run-on-all, I want to look at CVE-2020-28494 before doing a run).

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 1]
```diff
@@ -70,6 +70,61 @@ private DataFlow::Node getAValueExportedByPackage() {
...
+  // Object.defineProperties
+  exists(DataFlow::MethodCallNode call |
+    call = DataFlow::globalVarRef("Object").getAMethodCall("defineProperties") and
+    [call, call.getArgument(0)] = getAValueExportedByPackage() and
+    result = call.getArgument(any(int i | i > 0))
+  )
+  or
+  // Object.defineProperty
+  exists(DataFlow::MethodCallNode call |
+    call = DataFlow::globalVarRef("Object").getAMethodCall("defineProperty") and
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:87]: Can we use `CallToObjectDefineProperty` instead?

https://github.com/github/codeql/blob/746e9948b07f730c8c0a89e6ebcac5fa300b9d30/javascript/ql/src/semmle/javascript/StandardLibrary.qll#L5-L8

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 2]
```diff
@@ -70,6 +70,61 @@ private DataFlow::Node getAValueExportedByPackage() {
...
+    or
+    // assuming that the receiver of the call is somehow exported
+    result = map.getReceiver()
+  )
+  or
+  // Object.{fromEntries, freeze, entries, values}
+  exists(DataFlow::MethodCallNode freeze |
+    freeze =
+      DataFlow::globalVarRef("Object")
+          .getAMethodCall(["fromEntries", "freeze", "entries", "values"])
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:123]: We should also have `freezes` friend `seal` here.

(see https://github.com/github/codeql/blob/746e9948b07f730c8c0a89e6ebcac5fa300b9d30/javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll#L447)

**Review by esbena — CHANGES_REQUESTED** [review]
Can you make the `getAValueExportedByPackage` implementation use a bit more of our existing models?

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 3]
```diff
@@ -70,6 +70,61 @@ private DataFlow::Node getAValueExportedByPackage() {
...
+      call.getArgument(2)
+          .getALocalSource()
+          .getAPropertyReference("get")
+          .(DataFlow::FunctionNode)
+          .getAReturn()
+  )
+  or
+  // Object.assign
+  exists(DataFlow::MethodCallNode assign |
+    assign = DataFlow::globalVarRef("Object").getAMethodCall("assign")
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:102]: Huh. My comment about `ExtendCall` got lost.

Another case of potential reuse: https://github.com/github/codeql/blob/746e9948b07f730c8c0a89e6ebcac5fa300b9d30/javascript/ql/src/semmle/javascript/Extend.qll#L11

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 4]
```diff
@@ -70,6 +70,58 @@ private DataFlow::Node getAValueExportedByPackage() {
...
+    map.getMethodName() = ["map", "reduce", "entries", "values"] and
+    map = getAValueExportedByPackage()
+  |
+    result = map.getArgument(0).getABoundFunctionValue(_).getAReturn()
+    or
+    // assuming that the receiver of the call is somehow exported
+    result = map.getReceiver()
+  )
+  or
+  // Object.{fromEntries, freeze, entries, values}
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:116]:
```suggestion
  // Object.{fromEntries, freeze, seal, entries, values}
```

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 5]
```diff
@@ -70,6 +70,58 @@ private DataFlow::Node getAValueExportedByPackage() {
...
+    or
+    result =
+      call.getPropertyDescriptor()
+          .getALocalSource()
+          .getAPropertyReference("get")
+          .(DataFlow::FunctionNode)
+          .getAReturn()
+  )
+  or
+  // Object.assign
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:99]:
```suggestion
  // Object.assign and friends
```

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 6]
```diff
@@ -70,6 +70,58 @@ private DataFlow::Node getAValueExportedByPackage() {
     result = cla.getAStaticMethod() or
     result = cla.getConstructor()
   )
+  or
+  // *****
+  // Various standard library methods for transforming exported objects.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:75]:
```suggestion
  // Common styles of transforming exported objects.
```

File: javascript/ql/src/semmle/javascript/PackageExports.qll [hunk 7]
```diff
@@ -42,7 +42,7 @@ PackageJSON getTopmostPackageJSON() {
  * Gets a value exported by the main module from one of the topmost `package.json` files (see `getTopmostPackageJSON`).
  * The value is either directly the `module.exports` value, a nested property of `module.exports`, or a method on an exported class.
  */
-private DataFlow::Node getAValueExportedByPackage() {
+DataFlow::Node getAValueExportedByPackage() {
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/PackageExports.qll:45]:
```suggestion
private DataFlow::Node getAValueExportedByPackage() {
```
Do we have an uncommitted test file that needs this to be public?

**Review by esbena — COMMENTED** [review]
Thanks.
A few non-semantic nits.

File: javascript/ql/src/semmle/javascript/PackageExports.qll (hunk 7 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/PackageExports.qll:45]: No. I just accidentally committed something I shouldn't have.

**Review by esbena — APPROVED** [review]

**Comment by erik-krogh** [comment]
I've done a run-on-all with this, #5439, and #5458.  

The FP rate looks OK.  
We do flag some sub-packages that are only used internally in the monorepo.  
But we also flag a bunch of sub-packages that are used externally.  
And I don't see a way of distinguishing between the two. 

So I'll start merging.

**Pull request merged**